### ISSUE TITLE
[back] ci: use Black code style

### ISSUE THREAD

**Issue by GresilleSiffle** [tournesol-app/tournesol#471]
How about using `black` to format the back end code?

Currently we have several tools that check the back end code quality, but only `isort` is able to automatically format the code. `black` is able to format not only the import statements but the full files. This could help developpers to focus on new features and delegate the formatting to black.

The Tournesol back end will then share the same formatting as many other Python project, making it more " standard " and possibly more welcoming.

The `black` documentation explains how to make our current tools `isort`, `pylint` and `flake8` compatible with it:
https://black.readthedocs.io/en/stable/guides/using_black_with_other_tools.html

As @amatissart said, we might want to remove some of these tools if they bring nothing more than what `black` already brings.

**Comment by Vikka** [comment]
What follow is a reply from me to an issue of black on my code. (it was from Flake8 but I was told that it was black (it wasn't, but its not the point).

> Black is a PEP 8 compliant opinionated formatter.

It is not because of bugs. 

I checked the documentation of Black, and it appears that it should line break before the `or`.
https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#line-breaks-binary-operators

> Black will break a line before a binary operator when splitting a block of code over multiple lines. This is so that Black is compliant with the recent changes in the PEP 8 style guide, which emphasizes that this approach improves readability.

We can see that there is an issue from black:
https://github.com/psf/black/issues/1254
https://github.com/psf/black/issues/1781

I strongly suggest deactivating black for this kind of stuff error as it lower readability and make the code pep 8 uncompliant.

How to deactivate formatting for one line 
```python
foo = 1  # fmt: skip
```
https://github.com/psf/black/issues/451#issuecomment-1054372101


 and on multiple lines if desired: 
```python
# fmt: off
np.array(
    [
        [1, 0, 0, 0],
        [0, -1, 0, 0],
        [0, 0, 1, 0],
        [0, 0, 0, -1],
    ]
)
# fmt: on
```
https://stackoverflow.com/a/58584557/6251742

If this tool works and are acceptable to prevent black doing bad stuff, I think it will be a good idea.

An other issue is users relying on black 100%, which will lead to pep8 compliant (mostly) code, but not for human (by cutting line where it create ugly code, for example).

**Comment by Vikka** [comment]
https://github.com/tournesol-app/tournesol/pull/567#discussion_r835534740

This is the kind of stuff Black create: Black over Human. 
This is why I don't use black and automatic stuff. This lower the readability of the code for "convenience".
I think we should accept nor promote that kind of thing.

**Comment by Vikka** [comment]
[21:36] Oizo: 
>This is why I don't use black and automatic stuff. This lower the readability of the code for "convenience" of not disabling it when needed.

you may use a lot of energy and time to re-style a lot of code in the back end that black has already formatted, I think it could be a time saver (for you and the reviewers) to wait the deliberation about back / not black / custom black before continuing
[21:36] Oizo: readability matters, and at some point it becomes subjective
[21:37] Oizo: developers using black on a daily basis are accustomed to its style
[21:37] Oizo: i wasn't until i worked on Tournesol
[21:37] Oizo: black was new to me, and sometimes ugly
[21:37] Oizo: but the more I read it, the more I'm getting used to it
[21:38] Oizo: so.... I think we should wait the deliberation before re-writing more code
[21:39] Oizo: what's important in Tournesol? -> the public dataset ; the healthy recommendations ; the users privacy ; the " good " we can create in the world
[21:40] Oizo: taking care of the code base too, to make it accessible, easy to use and maintain
[21:40] Oizo: but yes, it's a matter of choice
[21:41] Oizo: ( and short-term priority at this moment )
[21:41] jst: Enforcing a linter with a common configuration also prevents commits touching a zillions of files just to lint according to author's config.
[21:42] Oizo: ( i like the word zillions )
[21:47] Vikka - FR: This enforce black style, not python style. Yes, it's a choice and I disagree with it 🙂 Lets put this in the issue discussion to document the choice for later.
[21:49] Vikka - FR: IMO, linter should helps, not drive. When it create ugly code, we should say stop, correct, and push it without worrying about later auto formating. This is why, IMO again, pylint is way better then black: advices, not skynet wanting to kill all humans

**Comment by Vikka** [comment]
This is how I see the dilema: 
1. black automated style everywhere, no compromisen even if it create ugly things,
2. black automated but fix ugly stuff and #fmtp: skip | off | on to prevent black "reformating", 
3. no black, pylint and stuff to warn about formating issues.

Here is the consequences I see: 
1. Black enpowerment, black style, non pep8 compliancy, non pythonic code, "I code this because I can read it, black broke it" stuff.
2. Will require a lot of comments in order to prevent black from linting long line in an ugly maner. Good maner teaching at a heavy cost on code.
3. Will require some work to teach how to read linter logs, or log cleaning to make useful stuff more obvious like a non-verbose option. Dev enpowerment.

What do you think ?

**Comment by Vikka** [comment]
## We must know why do we use our tools.

Black is a great tool to make a lot of thing in a minimum time. It's not a scalpel, it's a machinegun.
Useful to make a lot of work done quick, it struggle when it comes to edge cases where it can create code not really human friendly.

So let's do this: Why do we want to use Black?
- To gain time.
I agree, but what time ?
  - Time to make the code pep8 compliant.
  I agree, PEP8 is important, but why is it important?
    - Because if everybody is used to the same style, then everybody can work on everybody's code!
    Exactly! So we use black to make everybody able to read our code, quick!

### So, why do we want to use black? Make the code readable by everybody, fast (pep8 guidelines, code FOR HUMANS).

Black is useful when you don't want to spend time doing "obvious" things, but struggle in cases when you must [use your best judgment](https://peps.python.org/pep-0008/#a-foolish-consistency-is-the-hobgoblin-of-little-minds). 

Sometimes, PEP8 rules make the code harder to read. Sometimes, Black make the code harder to read. 

## "Yes, but we get used to it". 
Of course, we all can. With some effort, even the less used coding style can be learned. 
But this make the project harder to contribute, harder to understand, harder to improve.

The goal of style is to make the codebase more accessible. With automatic black, it became **less** accessible. 

## Examples:
Let's take an example:

```python
def f(table_A, object_B, object_C):
    table_A.object.filter(criteria=[data.progress() for data in object_B]).exclude(criteria=[data for data in object_C if data.progress() == 'done'])
```

Yes, pretty long line, isn't it? So let's black it.

```python
def f(table_A, object_B, object_C):
    table_A.object.filter(criteria=[data.progress() for data in object_B]).exclude(
        criteria=[data for data in object_C if data.progress() == "done"]
    )
```

Perfect, now, it's PEP8 compliant !!!! Or is it? PEP8 goal is to "improve the readability of code", as PEP20 zen of python says: Readability counts. So, is black achieve to make it pep 8 compliant? On paper, yes, in practice... This isn't a lot more readable.

So let's try it, as human:

```python
def f(table_A, object_B, object_C):
    table_A.object\
        .filter(criteria=[data.progress() for data in object_B])\
        .exclude(criteria=[data for data in object_C
                           if data.progress() == 'done'])
```

Great, so this is better! And it follow PEP8 guidelines too !

# But!

let run black again on this code...
```python
def f(table_A, object_B, object_C):
    table_A.object.filter(criteria=[data.progress() for data in object_B]).exclude(
        criteria=[data for data in object_C if data.progress() == "done"]
    )
```

In fact, black make the code less readable. Even when it's follow PEP8 and are really great.

## That's why it's a great tool, but should be carelessly applied on all the codebase.

# Let's take an other example, from PEP8 itself !

From [PEP8 - Maximum Line Length](https://peps.python.org/pep-0008/#maximum-line-length)
```python
with open('/path/to/some/file/you/want/to/read') as file_1, \
     open('/path/to/some/file/being/written', 'w') as file_2:
    file_2.write(file_1.read())
```

This code is great, very readable and shouldn't be refactored by black, isn't it?

```python
with open("/path/to/some/file/you/want/to/read") as file_1, open(
    "/path/to/some/file/being/written", "w"
) as file_2:
    file_2.write(file_1.read())
```

Don't. Use. Black. Take time to make the code great. 

> You will save time and mental energy for more important matters.

Style IS important. This is why you should care about your code. Don't let this stupid tool suck zen out of your code.

# Post-Scriptum:

## Black is great. 
Yes, I know, I look a bit bipolar. I use black. Or kinda. I have a tool like black. With the same downsides and upsides. 
## But
I choose **when** and **where** I run this tool. I can choose the line, the bloc, the function or class where I want it to run, without touching anything else. And when I do, I ALWAYS check if it's good. If not, ctrl+Z. If it's good but can be improved, I fix it. But I choose when I want it, and where. 

Like chainsaws, you don't want to use it EVERYWHERE EVERYWHEN. 
![image](https://user-images.githubusercontent.com/9381120/160281702-2a6949f6-1a2c-4a97-a131-c100fe97293a.png)

**Comment by Vikka** [comment]
If you struggle with python style, maybe you are using the wrong tools. Why use black if you don't need to?

My tool: 
![image](https://user-images.githubusercontent.com/9381120/160282072-87fe9dd0-ebdb-4619-b475-dea69a9a76da.png)

Great, I don't have to **know** pep8 ! Fixing a code "au fil de l'eau" is way easier than fixing it in the end.

Use modern and **configured** IDE to help you write python code if you don't know pep8 from your toes to your hairs.

This way, your code is good at the moment you write it, and this will help you during all the development process.

**Comment by amatissart** [comment]
About this issue it's quite interesting to read the proposal that led the Django project to recently format their code with Black:
https://github.com/django/deps/blob/main/final/0008-black.rst

It covers most of the arguments that have been discussed here.

These items from the "Motivation" section seems particularly relevant to me:

> With automated code formatting, this workload can be transferred to computers, saving human bandwidth for higher value-added activities. Generally speaking, there are two major benefits to automated code formatting:
>
>   *  Individually, developers no longer have to constantly make small decisions about code formatting, based on rules, on taste, or on a mix of both. Just throw the code in a file, run the formatter, done!  
>
>        You don't realize how much this reduces decision fatigue until you try it. It's a real boon to those, like me, who spend way too much time fine-tuning the formatting of their code.
>
 >  * Collectively, contributors no longer need to think about code formatting in code reviews. Whatever the formatter does is the expected result, by definition.
>
>    This removes the thought stream of "mmm, I would have formatted this a bit differently" while reviewing code exactly like it removes "mmm, how do I format this?" while writing code.

This resonates in me as I find discussing this kind of questions, with inherently subjective arguments, quickly exhausting.

To put it shortly, in my opinion, code style _is not so_ important. Readability is important, in some extent. (and enforcing consistency with automated formatting can certainly help with that, especially if we are using tools gaining popularity in the Python and Django communities).

And much more importantly, Tournesol is still quite a young project. There are many uncertainties about the directions it will take in the coming months ; any tool that can save us time and effort to focus on the big picture is worth taking.

**Comment by Vikka** [comment]
Great, use black. More, encourage it. But on developper side, not on server automated side.

Black can do most of the work. But in the end, let the developper IMPROVE its output. 

This is how I picture it.

1. When black is used before commit and push, it **can** be improved. It's a minimum level of quality.
2. When black is used after commit and push, it **can't** be improved. It's a maximum level of quality.

The 1. is why I (even if I fougth against) use black. To help me. But I know the limitation of the tool and often need to fix the edge case where black is **bad**. I focus on the important part, I don't let black take care of this parts for me, even when they already are greats.

Again: **USE BLACK**, but with caution, before commit, before PR, etc. 

Also, sytle IS important, since this is Open-Source. The code must follow standards and clear. For edge cases, black make it less readable, and this is why I am strongly against automated server side black.

Again, chainsaw are very useful, but not always.

(and you should'nt compare your projects to big one like Django. You don't have the same issues, and never will.)

**Comment by Vikka** [comment]
Of course, you can reply : "But if you don't want black to reformat your code, just use `# fmtp: on/off`. 

Great idea, then, instead of focusing only on edgecases, I'll also have to think about Black doing it's stuff, and how I will handle it. Instead of reducing mental energy required for formatting, it require more.

As I show, black increase consistency at the cost of readability in general. So, black is good, but 

### don't let it make the final choice.

**Comment by amatissart** [comment]
> Again, chainsaw are very useful, but not always.
>
>(and you should'nt compare your projects to big one like Django. You don't have the same issues, and never will.)

I totally agree on that. That's why we are looking for a solution that will fit our situation.

> Of course, you can reply : "But if you don't want black to reformat your code, just use # fmtp: on/off.

It wouldn't solve anything indeed. We can agree on that too. 

> Great, use black. More, encourage it. But on developper side, not on server automated side.

It does not seem very practical either, especially for newcomers. It means that applying Black would potentially reformat parts of the code that had been introduced in the same files by contributors who don't use black, although this code is unrelated to the changes to be committed, which would be counterproductive.

Overall, it seems that the remaining choices are options 1 and 3, as you described above:

>    **1.** black automated style everywhere, no compromise even if it create ugly things,
>    **3.** no black, pylint and stuff to warn about formating issues.

Option 1. would clearly be enforced by CI.  
What about option 3.? Would the CI enforce the linter checks? Any configuration example to bring more consistency while allowing some flexibility for developers?

**Comment by Vikka** [comment]
> It does not seem very practical either

Depends. Yes, because of the lack of configuration and the philosophy of Black (uncompromising formatter), it a lot "All Black, or nothing". I don't like that.

> It means that applying Black would potentially reformat parts of the code that had been introduced in the same files by contributors who don't use black, although this code is unrelated to the changes to be committed, which would be counterproductive.

Yes, totally, and this will make the job more difficult to newcomers. I don't like that either. I have a tool in mind, but even if it's very simple to use, it can look scary, will see about that later.

The issue with linter enforcing is that sometimes (less often than black), they don't like readable code because edge cases. This is an issue that can happen 1 time per 100-1000 lines, but it happens. If the linter run on all the code, it will trigger alert at the first edgecase and will again at each test if we don't accept this. Deactivate linter for the line can be acceptable but very boring to do.

## What I think is the best for newcomers: PyCharm, community version. 

Free, open-source, all in one tool that package a linter that show pep8 issues when you write it. If people don't want or struggle to run linter each time they push, this *migth* be a good thing for them. 

### Example
![image](https://user-images.githubusercontent.com/9381120/160346696-732a9978-0b55-4c67-b127-307676c6788f.png)
Here, the tool show where there is a lint warning (yellow underline).

When the user move its cursor or mouse on the warning, she can see the warning: 
![image](https://user-images.githubusercontent.com/9381120/160346997-9f70902c-4023-4aad-bf1a-c92106e69c51.png)

Also great to warn about dangerous practices:
![image](https://user-images.githubusercontent.com/9381120/160347094-da15a38b-239a-4d78-aab2-4f9f355b6e65.png)
And provide some help about python:
![image](https://user-images.githubusercontent.com/9381120/160347142-4f418429-bc97-4798-a99e-2d525e0480bf.png)

Without preventing the user to commit this code, it will warn the user that there is a mistake when she write it. I use this feature to teach Python Style to my students, I found this feature great and efficient when it comes to teach good practices in python.

Of course, the tool package a formatter, nicer than black, but very useful. You can run is on a selected block of code to fix a lot of issues in 1 sec:
![image](https://user-images.githubusercontent.com/9381120/160347801-7b555a20-7679-4fd1-a977-fbb7e6524d06.png)
Run the formatter:
![image](https://user-images.githubusercontent.com/9381120/160347864-de73dc34-5618-4029-9c7c-9e7946cca755.png)

**Comment by GresilleSiffle** [comment]
Hélo everyone,

We will take schedule a proper meeting to discuss about `black` and make a proposition.

More over it seems important that at least every member of the core team can share their ideas and opinions, as some of use are working full time on the code base and the decision will impact our daily works.

If there is no objection, I'd like to propose to:
- use the " Gestion par consetement " design to decide (exposition / ideas / proposition / feeling / objections etc.)
- this design emphasis the fact that everyone will have the same amount of time to speak
- keep track of the decision in our Architecture Decision Log so that if the subject appears again, we will be able to easily share the why of our decision

Let's do this after the french election dead line : )

**Comment by Vikka** [comment]
Any update about this issue?
